igvc_perception/src/pointcloud_filter/fast_segment_filter/fast_segment_filter.cpp

```diff
@@ -0,0 +1,317 @@
+#include <pointcloud_filter/fast_segment_filter/fast_segment_filter.h>
+
+namespace pointcloud_filter
+{
+void FastSegmentFilter::filter(pointcloud_filter::Bundle &bundle)
+{
+  segments_.clear();
+  ground_points_.clear();
+  nonground_points_.clear();
+
+  for (const auto &point : bundle.pointcloud->points)
+  {
+    int segment_id = getAngleFromPoint(point) * config_.num_segments / 360;
+    segments_[segment_id].raw_points_.emplace_back(point);
+  }
+  processSegments();
+  getLinesFromSegments();
+  if (config_.debug_viz)
+  {
+    debugViz();
+  }
+  ground_points_.header.frame_id = "/lidar";
+  nonground_points_.header.frame_id = "/lidar";
+  classifyPoints(ground_points_, nonground_points_);
+  ground_pub_.publish(ground_points_);
+  nonground_pub_.publish(nonground_points_);
+
+  bundle.free_pointcloud->points = std::move(ground_points_.points);
+  bundle.occupied_pointcloud->points = std::move(nonground_points_.points);
+}
+
+double Line::distFromPoint(const Prototype point)
+{
+  Eigen::Vector3d vect;
+  vect(0) = point.point_.x;
+  vect(1) = point.point_.y;
+  vect(2) = point.point_.z;
+  double t = -(intercept_ - vect).dot(params_) / (params_.dot(params_));
+  Eigen::Vector3d pt_on_line = t * params_ + intercept_;
+  return (pt_on_line - vect).squaredNorm();
+}
+
+bool Line::fitPoints(const Prototype new_point)
+{
+  int total_points = model_points_.size() + 1;
+  if (total_points < 3)
+  {
+    if (model_points_.size() == 0)
+    {
+      start_point_ = new_point;
+    }
+    else
+    {
+      end_point_ = new_point;
+    }
+    model_points_.emplace_back(new_point);
+    return true;
+  }
+  Eigen::Vector3d old_params = params_;
+  Eigen::Vector3d old_intercept = intercept_;
+  MatrixXd A(total_points, 3);
+  MatrixXd mean(1, 3);
+  for (int i = 0; i < total_points; i++)
+  {
+    Prototype pt;
+    if (i == total_points - 1)
+    {
+      pt = new_point;
+    }
+    else
+    {
+      pt = model_points_[i];
+    }
+
+    A(i, 0) = pt.point_.x;
+    A(i, 1) = pt.point_.y;
+    A(i, 2) = pt.point_.z;
+    mean(0, 0) += pt.point_.x;
+    mean(0, 1) += pt.point_.y;
+    mean(0, 2) += pt.point_.z;
+  }
+  mean /= -total_points;
+  A += mean.replicate(total_points, 1);
+  Eigen::JacobiSVD<MatrixXd> svd(A, Eigen::ComputeFullU | Eigen::ComputeFullV);
+  params_ = svd.matrixV().col(0);
```

Comment:
Oh is `params_` just equal to the largest eigenvector? I think the member variable could be renamed to be a bit more clear.

Reply:
I renamed to slope_ to be more clear, and I added documentation for the Line struct explaining it.